### DISCUSSION TITLE
Browser based RNode flasher

### DISCUSSION THREAD

**Discussion by Erethon** [markqvist/Reticulum#471] [Show and tell]
Hello, I've setup an experimental browser based flasher to flash ESP32 devices with the RNode firmware. You can find this experiment in https://rnode-flasher.erethon.com/. I've tested this against two boards, the old (SX127X based) T-Beam and Lora T3 v1.6.1 and it works on both of them. I've experimentally also uploaded the firmware for the T3s with the TCXO, but I haven't tested myself if it works as expected.

I would like to ask the community for help with testing in different setups and boards. In order for it to work you'll need a browser that supports [webserial](https://developer.mozilla.org/en-US/docs/Web/API/Web_Serial_API) (practically this means Chromium/Chrome/Edge/Opera).  As mentioned, currently it only supports three boards (in both 868/915 and 433 MHz) but I'll add all the RNode firmware supported boards in the next few days.

Flashing the firmware is currently done using a slow speed to avoid some timing issues I was seeing initially, so it takes around 4-5 minutes to flash a board. This work is based on https://github.com/esphome/esp-web-tools, once things are a bit more stable I'll push my changes under https://github.com/Erethon/rnode-flasher which is a fork of `esp-web-tools`.

I'll be updating this discussion with ToDos as things progress.
#### TODO

- [ ] Add all boards supported by rnode firmware
- [ ] Verify flashing works as expected
  - [X] Old (SX127X) TBeam
  - [X] T3 V1.6.1
  - [x] T3 V1.6.1 with TCXO
  - [x] Heltec32 V2
  - [x] TBeam with SX126X
- [ ] Make it flash boards faster
- [ ] Add support for Bluetooth related settings (enable Bluetooth, pairing, etc)

The last checkbox about Bluetooth is an important one, as the flasher is gonna be mostly useful for people that want to use Reticulum over Sideband.

#### Feedback
Any kind of feedback from "it works" or "it doesn't work on X/Y/Z setup" (especially if you can provide the browser console logs as well) is highly welcome!

**Comment by markqvist** [comment]
It works great on T3 V1.6.1 based RNode, which is all I had with me right now. I'll try with more when I get a chance :)

Looking great so far! Thanks so much for the effort! Awesome stuff!

**Comment by faragher** [comment]
Tried the TCXO variant again, same hardware failure message. Reboots twice, as expected, but it doesn't boot correctly. I'm going to check the EEPROM contents if I can and see if that's where the problem is.

The other two boards function flawlessly, as I mentioned.

**Reply by Erethon** [reply]
An EEPROM dump will be helpful indeed, it's either gonna be all FFs which means it's a timing issue (or another form of failing to properly write to the serial port) or something silly on my side (setting a byte wrong for example).

I think rnodeconf won't allow you to do an eeprom dump if the checksum isn't correct (the all FFs case). A `RNS.log()` before [this if](https://github.com/markqvist/Reticulum/blob/master/RNS/Utilities/rnodeconf.py#L804) should get around this problem.

Thank you for your testing efforts!

**Reply by faragher** [reply]
Yes, it's an EEPROM content issue:

0x1b to 0x9a is uninitialized (128 bytes). Probably just didn't copy the signing key or something similar.

**Reply by faragher** [reply]
As requested, the first section (truncated because it's a lot of FFs followed by the lock byte)

Web flasher:
`b1:09:01:00:00:00:0b:65:fb:5d:a8:51:08:10:12:65:68:88:cc:09:b9:37:d7:2b:a0:e5:29`

RNodeConf:
`b1:b9:01:00:00:00:06:65:fb:5f:90:13:65:0f:d1:9f:a5:43:da:37:ae:06:60:e9:4f:d2:2f:70`

The model appears to be different, as do the serial, time, and checksum (as expected).  Probably the model has to be set to B9 instead of 09.

**Reply by Erethon** [reply]
Sounds simple enough to fix, I went by [this](https://github.com/markqvist/Reticulum/blob/master/RNS/Utilities/rnodeconf.py#L153) and [this](https://github.com/markqvist/Reticulum/blob/master/RNS/Utilities/rnodeconf.py#L1907), but it seems that we're not always using that (i.e. [here](https://github.com/markqvist/Reticulum/blob/master/RNS/Utilities/rnodeconf.py#L3051) @markqvist  do you know why this is the case?). Pushing an update for this now.

**Reply by faragher** [reply]
I think it's likely a typo in the first RNodeconf link, because I can't find a 0b model definition anywhere.

**Reply by Erethon** [reply]
I've updated the [landing page](https://rnode-flasher.erethon.com/) to also include two new entries marked "Temporary B9/B4 model tests" as a test to see if this fixes it.

**Reply by faragher** [reply]
The new TCXO version works for me now.

**Reply by markqvist** [reply]
Ooooh, yeah, that's confusing. The `0x04` and `0x04` model codes are basically a hack to get `rnodeconf` to differentiate the boards during the autoinstall procedure! Once it's selected the correct firmware, it [maps the actual model code back](https://github.com/markqvist/Reticulum/blob/master/RNS/Utilities/rnodeconf.py#L3049) to `0xB4` or `0xB9`.

I've added a comment in the source to avoid future confusion :)

**Comment by markqvist** [comment]
This is starting to look really great @Erethon!

We'd have pretty broad coverage, if we could also add support for the following board types:

- T-Beams based on SX1262/8
- Heltec LoRa32 V2 boards

I have access to these in both 868 and 433 MHz variants, and can test them, if you can add the definitions for them to the flasher.

**Reply by Erethon** [reply]
I've uploaded files for both but I can't test it myself unfortunately, let me know how it goes!

**Comment by markqvist** [comment]
Also, is it realistic to add Bluetooth controls to this utility, so users can easily enable Bluetooth and pairing mode straight from the flasher? I'm not sure if the architecture of the flasher script is amenable to something like that, or would it require a larger effort?

Otherwise, it might be a good idea to point out that this can be done from within the "Hardware" section of Sideband (available on Android only, though).

**Reply by Erethon** [reply]
Yes, I think it's realistic to do this via the flasher and actually a mandatory step to make it actually useful to people. It's already the last bullet on the TODOs in the original post.

**Reply by markqvist** [reply]
Yes, sorry for not being clear, I was just curious whether you thought it would be realistic to do it within a reasonable amount of effort, or if there was something underlying in the flasher script that would make it difficult to do.

**Comment by markqvist** [comment]
I can confirm that both the Heltec V2 boards and SX1262/8 T-Beams work fine with the web flasher now too :)

**Comment by SubBass100** [comment]
Hello, cool project! I have a (433) ESP32-S3 (T-Beam-Core S3) https://www.lilygo.cc/products/softrf-t-beamsupreme?variant=42880905052341 I'd be willing to run some tests for you If you are interested, let me know!

**Comment by markqvist** [comment]
In the [latest RNode firmware release](https://github.com/markqvist/RNode_Firmware/releases/tag/1.72), I added support for The LilyGO T3S3, and Heltec V3 boards (even though they are serial/USB only for now, no BLE yet). Can these be added to the web flasher too?

**Reply by cobraPA** [reply]
Along those lines, would it be possible to use this to flash custom firmware?  While Mark's version does not do BLE, the iOS RNode test area has iOS BLE support for the Heltec V3.

**Reply by Erethon** [reply]
I'm a little behind on updates that have happened in the firmware in the last few weeks. Once I get a bit more time (probably over the weekend or early next week), I'll look into adding this and upstreaming all my work to fully publish the web flasher.

**Reply by markqvist** [reply]
It would probably also be good to coordinate with @jacobeva to get the web flasher incorporated into the RNode Firmware CE repo!

**Reply by faragher** [reply]
We should also look into a standard suite of servers to provide RNode firmware to any given flasher from an arbitrary repo. In my usual way, I did something and then forgot about it. 😆 

https://github.com/faragher/RRepo

**Reply by jacobeva** [reply]
> It would probably also be good to coordinate with @jacobeva to get the web flasher incorporated into the RNode Firmware CE repo!

Can you explain what you mean by this Mark? Do you mean adding my repo as a firmware package source?

Thanks for that link Mike, I will look at your library there and see what I can do with it. I'm thinking of creating a custom version of rnodeconf anyway, so I may include support for that.

In terms of clearnet stuff, is there any source publicly available for a fw repo server? I do not know what unsigned.io is running, and I am unsure how to proceed to deploy my own.

**Reply by faragher** [reply]
I've run into an issue with Mark's server. It doesn't seem to respect requesting a specific firmware version. I don't remember exactly what he said, but my recollection is it was already something that needed retooling, so it may be best to start from zero. The API can be carried over for interoperability, at least as far as RNodeconf calls it. I seem to recall it's just a request with the question mark . . . I can never remember what that's called.

**Comment by jacobeva** [comment]
It's a URL parameter. Okay, I'll see what I can whip up...

**Comment by markqvist** [comment]
Just so you don't feel bound up by how it was done previously, feel free to completely retool how the update / firmware file acquisition works into something much better. My only wish it that it allows users to easily choose/add firmware repos themselves. The current system was designed a long time ago, when things were very different, and the whole thing only targeted two different board types, so there is a lot of carry-over from that.

In a similar vein, it might be a fine enough time to drop support for the old AVR platform from the CE version of the firmware. Would have posted about that somewhere else, but this will be as fine a place as any, I guess. There hasn't been any new features added to AVR based RNodes for a long time, and having to keep support for those boards, with very limited resources, will hinder implementation of more advanced features. And my repo will keep hosting files for the AVR-based RNodes already out there, and in the odd case someone builds one, anyway, so moving on from that here would probably make sense.

Also, maybe it's really worth considering a new RNode configuration/updater/flasher utility, something like `rnodetool` or whatever :) As it often happens, `rnodeconfig` started its life with some much more limited goals than what it has grown into today, and as a result has become a bit unwieldy in its codebase. It may be worth considering creating a new tool, scissoring out all the good bits from `rnodeconf`, organising and updating them accordingly, and setting everything up in an easier to manage and more future-oriented way.

Just wanted to put that out there, so we nobody feels we have to hang onto something just for the sake of it. Always worth considering if something is ripe for trimming and/or a bit of re-contextualisation.

**Reply by jacobeva** [reply]
Okay, once I have finished with my current work on the CE, I will get to work on a new flashing utility which supports fiat (sic) internet and RNS firmware repos. `rnodetool` is not a bad name :) It's good to know I have your consent to throw the old system out the window if need be, I probably will if it's too cumbersome (and it's a good time to do it given I've started the CE fork as well).

Yes, I will be removing support for AVR. You may have noticed it's not in the supported boards list for the repo, that was intentional. I will likely be doing that soon as a part of refactoring I am currently doing for adding multiple interface support.

**Comment by liamcottle** [comment]
For future reference, I've also been working on a web based RNode flasher:
- https://github.com/liamcottle/rnode-flasher
- https://liamcottle.github.io/rnode-flasher/
